java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java

```diff
@@ -48,10 +48,21 @@ public ByChained(By... bys) {
 
   @Override
   public WebElement findElement(SearchContext context) {
-    List<WebElement> elements = findElements(context);
-    if (elements.isEmpty())
-      throw new NoSuchElementException("Cannot locate an element using " + toString());
-    return elements.get(0);
+    if (bys.length == 0) {
+      throw new NoSuchElementException("No Bys were specified in this " + getClass().getSimpleName());
+    }
+
+    WebElement elem = null;
+    for (By by : bys) {
+      boolean isFirstIteration = elem == null;
+      if (isFirstIteration) {
+        elem = by.findElement(context);
+      } else {
+        elem = elem.findElement(by);
+      }
+    }
+    assert elem != null; // NoSuchElement should have already been thrown
```

Comment:
Hey
sorry, I'm back.
It looks like the current behavior is actually the way I thought.
Please check out https://github.com/alb-i986/selenium/commit/54d0dc8d72f2643183d7f5bbe96ff727fbae9f93

The key is ` elems = newElems;` at the end of each iteration, which overwrites the list that is iterated during each step. The first by that returns no elements will set `elems` to empty list; any subsequent iteration of the outer loop will do nothing, since the inner loop will not be executed. BTW this lead me to another improvement: to [exit the loop as soon as elems is empty](https://github.com/alb-i986/selenium/commit/ae639dab1b3a114ae5ad396bdcc315c4fc1d627c)

I made some more refactorings, including replacing the for-each with a standard for, which is supposed to improve readability.
https://github.com/alb-i986/selenium/blob/d336458e8040e7bf4cbc09afe1232777dbe8c0f1/java/client/src/org/openqa/selenium/support/pagefactory/ByChained.java#L74

Please let me know your thoughts
Thanks